invoke-ai/InvokeAI
# refactor: simplified config system

## What type of PR is this? (check all applicable)


## Have you discussed this change with the InvokeAI team?

      
## Have you updated all relevant documentation?
- Probably missed a few spots


## Description

I ran into some issues with interactions between argparse, omegaconf and pydantic validation + existing issues:
- argparse autogenerated CLI args for every config setting and this cannot handle structured config data (for example, the list of url regex & token pairs). It was not simple to just skip those settings due to the tight coupling of argparse and the config object.
- We were writing every default config setting to the config file. This is a longstanding issue. With this PR, that will no longer happen, and a simple migration from v3 config to v4 partially tidies all configs. I sneakily changed the name of `png_compress_level` to `pil_compress_level` - this will get everybody using the better default value of 1. 
- There was an awkward API with `InvokeAIAppConfig.get_config`.

It was all very intertwined and I couldn't figure out an easy fix, so I refactored the whole thing and simplified it substantially.

See the commit messages for details on the changes, especially 6b066f3b44af1cf079310c8b075abeea532e5899 which has the meat and potatoes. Most changes are just small changes to use the revised config system.

The API did change slightly from that commit, to support testing and the configurator.

## QA Instructions, Screenshots, Recordings

- Start up the app, your `invokeai.yaml` should change to the new flat format, and the old one backed up to `invokeai.yaml.bak`. Your YAML should now only contain settings that are different from defaults. Everything should work exactly the same as it did before - like custom paths for models, host, port.
- Restore your backed up YAML & delete the `.bak` file. Run the configurator - it should migrate the YAML as it starts and back it up, same as before. Change a couple settings and apply changes, the YAML should change accordingly.

It's hard to fully test this change because the config is used all over the place, but I was pretty darn careful when updating all the places the config is used.



## Merge Plan

This PR can be merged when approved



## Added/updated tests?

      have not been included_


Closes #5964
